internal/query/zedtokens.go

```diff
@@ -0,0 +1,145 @@
+package query
+
+import (
+	"context"
+	"errors"
+
+	pb "github.com/authzed/authzed-go/proto/authzed/api/v1"
+	"github.com/nats-io/nats.go"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+
+	"go.infratographer.com/permissions-api/internal/types"
+)
+
+const (
+	consistencyMinimizeLatency = "minimize_latency"
+	consistencyAtLeastAsFresh  = "at_least_as_fresh"
+)
+
+// getLatestZedToken attempts to get the latest ZedToken for the given resource ID.
+func (e *engine) getLatestZedToken(ctx context.Context, resourceID string) (string, error) {
+	_, span := e.tracer.Start(
+		ctx,
+		"getLatestZedToken",
+		trace.WithAttributes(
+			attribute.String(
+				"permissions.resource",
+				resourceID,
+			),
+		),
+	)
+
+	defer span.End()
+
+	resp, err := e.kv.Get(resourceID)
+	if err != nil {
+		// Only record this as an error if it wasn't a not found error.
+		if !errors.Is(err, nats.ErrKeyNotFound) {
+			span.RecordError(err)
+			span.SetStatus(codes.Error, err.Error())
+		}
+
+		return "", err
+	}
+
+	zedToken := string(resp.Value())
+
+	return zedToken, nil
+}
+
+// upsertZedToken updates the ZedToken at the given resource ID key with the provided ZedToken.
+func (e *engine) upsertZedToken(ctx context.Context, resourceID string, zedToken string) error {
+	_, span := e.tracer.Start(
+		ctx,
+		"upsertZedToken",
+		trace.WithAttributes(
+			attribute.String(
+				"permissions.resource",
+				resourceID,
+			),
+		),
+	)
+
+	defer span.End()
+
+	zedTokenBytes := []byte(zedToken)
+
+	// Attempt to get a ZedToken. If we found one, update it. If not, create it. If some other error
+	// happened, log that and return
+	resp, getErr := e.kv.Get(resourceID)
+
+	var err error
+
+	switch {
+	// If we found a token, update it. This may fail if another client updated it before we did
+	case getErr == nil:
+		_, err = e.kv.Update(resourceID, zedTokenBytes, resp.Revision())
+	// If we did not find a token, create it. This may fail if another client created an entry already
+	case errors.Is(getErr, nats.ErrKeyNotFound):
+		_, err = e.kv.Create(resourceID, zedTokenBytes)
+	// If something else happened, just keep moving
+	default:
+	}
+
+	// If an error happened when creating or updating the token, record it.
+	if err != nil {
```

Comment:
ill be interested to see how often we hit this condition - i expect not a lot since its for an individual relationship

Reply:
Probably not often, but I can think of a few cases:
* NATS is unavailable
* Something created a bucket entry before this function did
* Something updated a bucket entry